packages/components/src/utils/hooks/use-focus-outside/index.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * External dependencies
+ */
+import { includes } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Input types which are classified as button types, for use in considering
+ * whether element is a (focus-normalized) button.
+ *
+ * @type {string[]}
+ */
+const INPUT_BUTTON_TYPES = [ 'button', 'submit' ];
+
+/**
+ * Returns true if the given element is a button element subject to focus
+ * normalization, or false otherwise.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+ *
+ * @param {HTMLElement} element Element to test.
+ *
+ * @return {boolean} Whether element is a button.
+ */
+function isFocusNormalizedButton( element ) {
+	switch ( element.nodeName ) {
+		case 'A':
+		case 'BUTTON':
+			return true;
+
+		case 'INPUT':
+			return includes(
+				INPUT_BUTTON_TYPES,
+				/** @type {HTMLInputElement} */ ( element ).type
+			);
+	}
+
+	return false;
+}
+
+/**
+ *
+ * @param {Function} onFocusOutside
+ * @param {import('react').Ref<HTMLElement>} wrapperRef
+ */
+export default function useFocusOutside( onFocusOutside, wrapperRef ) {
+	const ref = useRef( wrapperRef );
+	const preventBlurCheck = useRef( false );
+
+	/**
+	 * @type {import('react').MutableRefObject<number | undefined>}
+	 */
+	const blurCheckTimeoutId = useRef();
+
+	const cancelBlurCheck = () => {
+		clearTimeout( blurCheckTimeoutId.current );
+	};
+
+	useEffect( () => {
+		cancelBlurCheck();
+	}, [ cancelBlurCheck ] );
+
+	/**
+	 * Handles a mousedown or mouseup event to respectively assign and
+	 * unassign a flag for preventing blur check on button elements. Some
+	 * browsers, namely Firefox and Safari, do not emit a focus event on
+	 * button elements when clicked, while others do. The logic here
+	 * intends to normalize this as treating click on buttons as focus.
+	 *
+	 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+	 *
+	 * @param {import('react').SyntheticEvent} event Event for mousedown or mouseup.
+	 */
+	const normalizeButtonFocus = ( event ) => {
+		const { type, target } = event;
+		const isInteractionEnd = includes( [ 'mouseup', 'touchend' ], type );
+
+		if ( isInteractionEnd ) {
+			preventBlurCheck.current = false;
+		} else if (
+			isFocusNormalizedButton( /** @type {HTMLElement} */ ( target ) )
```

Comment:
Should this be [`currentTarget`](https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget)?

> The **`currentTarget`** read-only property of the `Event` interface identifies the current target for the event, as the event traverses the DOM. It always refers to the element to which the event handler has been attached, as opposed to `Event.target`, which identifies the element on which the event occurred and which may be its descendant.

_Take this with a grain of salt, I'm not especially familiar with how this is intended to be used._

The default type arguments for [`SyntheticEvent`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c1bd87eb4005c5fd8866e90d66d851583aa58ae0/types/react/index.d.ts#L1141-L1166) suggest this, where `currentTarget` defaults to type [`Element`](https://github.com/microsoft/TypeScript/blob/ee570402769c3392d82a746fdf1416e4ce96304d/lib/lib.dom.d.ts#L5098) (which extends [`Node`](https://github.com/microsoft/TypeScript/blob/ee570402769c3392d82a746fdf1416e4ce96304d/lib/lib.dom.d.ts#L10752)). Synthetic event (with default type arguments) here resolves to something like:

```ts
interface SyntheticEvent {
  currentTarget: EventTarget & Element;
  target: EventTarget;
  // …
}
```

Note that `SyntheticEvent` doesn't give us access to pass parameters to `target`, which suggests that we can't reason much about it.

Reply:
I always have to look at this code closely to re-understand it!

I don't think it should be `currentTarget`, as mentioned that would be the element the handler is bound to (usually some wrapping `div`), but what this function is trying to determine is whether a child input, link or button had some interaction.

Reply:
Nice, that's really helpful 👍 

I took a look at why these types might be like this and found this MDN on [`EventTarget`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget):

> `Element`, `Document`, and `Window` are the most common event targets, but other objects can be event targets, too. For example `XMLHttpRequest`, `AudioNode`, `AudioContext`, and others.

It seems like the safest thing to do may be to make `isFocusNormalizedButton` accept an `EventTarget`.

<details>

<summary>Diff to accept `EventTarget`</summary>

```patch
diff --git a/packages/components/src/utils/hooks/use-focus-outside/index.js b/packages/components/src/utils/hooks/use-focus-outside/index.js
index b04efbf10d..690acd0a15 100644
--- a/packages/components/src/utils/hooks/use-focus-outside/index.js
+++ b/packages/components/src/utils/hooks/use-focus-outside/index.js
@@ -16,18 +16,22 @@ import { useEffect, useRef } from '@wordpress/element';
  */
 const INPUT_BUTTON_TYPES = [ 'button', 'submit' ];
 
+/* eslint-disable jsdoc/valid-types */
 /**
  * Returns true if the given element is a button element subject to focus
  * normalization, or false otherwise.
  *
  * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
  *
- * @param {HTMLElement} element Element to test.
+ * @param {EventTarget} eventTarget Element to test.
  *
- * @return {boolean} Whether element is a button.
+ * @return {EventTarget is HTMLElement} Whether element is a button.
  */
-function isFocusNormalizedButton( element ) {
-	switch ( element.nodeName ) {
+function isFocusNormalizedButton( eventTarget ) {
+	if ( ! ( eventTarget instanceof window.HTMLElement ) ) {
+		return false;
+	}
+	switch ( eventTarget.nodeName ) {
 		case 'A':
 		case 'BUTTON':
 			return true;
@@ -35,12 +39,13 @@ function isFocusNormalizedButton( element ) {
 		case 'INPUT':
 			return includes(
 				INPUT_BUTTON_TYPES,
-				/** @type {HTMLInputElement} */ ( element ).type
+				/** @type {HTMLInputElement} */ ( eventTarget ).type
 			);
 	}
 
 	return false;
 }
+/* eslint-enable jsdoc/valid-types */
 
 /**
  * @typedef {import('react').SyntheticEvent} SyntheticEvent
@@ -102,9 +107,7 @@ export default function useFocusOutside( onFocusOutside, __unstableNodeRef ) {
 
 		if ( isInteractionEnd ) {
 			preventBlurCheck.current = false;
-		} else if (
-			isFocusNormalizedButton( /** @type {HTMLElement} */ ( target ) )
-		) {
+		} else if ( isFocusNormalizedButton( target ) ) {
 			preventBlurCheck.current = true;
 		}
 	};
```

Reply:
Awesome, thanks for the suggestion, that does feel better!

Does something like the following also work for the return value of `isFocusNormalizedButton`?
```jsx
/**
 * @typedef {HTMLButtonElement | HTMLLinkElement | HTMLInputElement} FocusNormalizedButton
 *
 * // ...
 *
 * @return {eventTarget is FocusNormalizedButton} Whether element is a button.
 */
```

Was wondering if that'd be more accurate (since the return could also be a link or a button)?

Reply:
Nice work 😎  That's more accurate, but it may not be very different because we'll only be able to use the common parts of that type union. For example, you can access `href` on something of type `HTMLLinkElement`, but not `HTMLLinkElement | HTMLButtonElement` because `href` doesn't exist on `HTMLButtonElement`.

👍  This bit is to go either way and should be safer.